### ISSUE TITLE
Allow anonymous functions in Style

### ISSUE THREAD

**Issue by keenancrane** [penrose/penrose#263]
Currently all `encourage` and `ensure` statements in Style must be associated with a named function.  For instance, here's a snippet from the Venn diagram style:

```
Set x {
    x.shape = Circle {
        strokeWidth : 1
    }

    x.text = Text {
        string : x.label
    }

    x.labelFn = ensure contains(x.shape, x.text)
    x.minSizeFn = ensure minSize(x.shape)
    x.maxSizeFn = ensure maxSize(x.shape)
    x.labelPosFn = ensure outsideOf(x.text, x.shape)
    LOCAL.layering  = x.shape below x.text
}
```

While it is certainly convenient to be able to refer to the functions `labelFn`, `minSizeFn`, etc., it also makes code unnecessarily verbose.  In this case, for instance **none** of these functions are ever referred to later on.  Ideally, the code would just look like this:

```
Set x {
    x.shape = Circle {
        strokeWidth : 1
    }

    x.text = Text {
        string : x.label
    }

    ensure contains(x.shape, x.text)
    ensure minSize(x.shape)
    ensure maxSize(x.shape)
    ensure outsideOf(x.text, x.shape)

    LOCAL.layering  = x.shape below x.text
}
```

This code reads much more naturally (which is after all the purpose of having a domain-specific language!): _"Please just ensure that a set of things happen."_  Currently, one must instead invent a bunch of names that are never used, and serve only to clutter the code.

This would also help reduce use of the (IMO) awful "LOCAL" syntax.  For instance, this code

```
Set x; Set y
where IsSubset(x, y) {
    LOCAL.containFn = ensure contains(y.shape, x.shape)
    LOCAL.sizeFn    = ensure smallerThan(x.shape, y.shape)
    LOCAL.outsideFn = ensure outsideOf(y.text, x.shape)
    LOCAL.layering  = x.shape above y.shape
    LOCAL.layering1  = y.text below x.shape
}
```

would become just

```
Set x; Set y
where IsSubset(x, y) {

    ensure contains(y.shape, x.shape)
    ensure smallerThan(x.shape, y.shape)
    ensure outsideOf(y.text, x.shape)

    LOCAL.layering  = x.shape above y.shape
    LOCAL.layering1  = y.text below x.shape
}
```

**Comment by keenancrane** [comment]
One way to implement this feature, perhaps, is to just treat it as sugar: if the compiler encounters a bare `ensure` or `encourage` keyword, then it can just expand it into `X.enc1 = encourage`, `X.enc2= encourage`, `X.ens1 = ensure`, etc., where the function names are generated programmatically, and `X` is determined from the local scope.

**Comment by hypotext** [comment]
Thanks @keenancrane, I think this language feature can be treated together with #264. I agree that it would make the user experience nicer. 

The question to answer is: What Substance object should a optimization function or layering statement belong to if there are multiple objects in scope? One of them? None (use LOCAL)? The compiler would need to decide this, for example in this case: 

```
Set x; Set y
where IsSubset(x, y) {
    ensure contains(y.shape, x.shape)
    x.shape above y.shape
}
```

@wodeni Let's talk about how to parse this, and I can deal with it on the compiler side.

**Comment by hypotext** [comment]
Resolved in #317
